File: 编程/leetcode/1475-[E]-商品折扣后的最终价格/solution.py
from typing import List


class Solution:
    def finalPrices(self, prices: List[int]) -> List[int]:
        stack = []
        for idx, p in enumerate(prices):
            while stack and p <= prices[stack[-1]]:
                prices[stack[-1]] = prices[stack[-1]]-p
                stack.pop()
            stack.append(idx)
        return prices


if __name__ == '__main__':
    test = Solution()
    print(test.finalPrices([8,4,6,2,3]))
    print(test.finalPrices([1,2,3,4,5]))
    print(test.finalPrices([10,1,1,6]))
        

File: 编程/leetcode/1475-[E]-商品折扣后的最终价格/solution.rs
struct Solution;

impl Solution {
    pub fn final_prices(prices: Vec<i32>) -> Vec<i32> {
        let mut stack =  Vec::new();
        let mut new_price = Vec::new();
        for (idx, &p) in prices.iter().enumerate() {
            new_price.push(p);
            while !stack.is_empty() && p <= prices[stack[stack.len()-1] as usize] {
                new_price[stack[stack.len()-1] as usize] -= p;
                stack.pop();
            }
            stack.push(idx as i32);
        };
        return new_price;
    }
}

fn main() {
    println!("{:?}", Solution::final_prices(vec![8,4,6,2,3]));
    println!("{:?}", Solution::final_prices(vec![1,2,3,4,5]));
}